pkg/newlog/cmd/newlogd.go

```diff
@@ -592,7 +592,10 @@ func getMemlogMsg(logChan chan inputEntry, panicFileChan chan []byte) {
 			if logInfo2.Source == "" {
 				logInfo2.Source = logInfo.Source
 			}
+			// Replace the JSON struct with the inner JSON struct
 			logInfo = logInfo2
+			// but keep the original message text and fields
+			logInfo.Msg = logEntry.Msg
```

Comment:
This works, but it would match the pattern on line 590,593 if you set it from logInfo.Msg instead of logEntry.Msg.
Or perhaps this can be structures in a better way?

Also, since we've failed to have tests for this, it is possible to do some newlogd unit tests for this which would extercise the various combinations of inputs?

Reply:
> This works, but it would match the pattern on line 590,593 if you set it from logInfo.Msg instead of logEntry.Msg.
> Or perhaps this can be structures in a better way?

Sorry, I don't understand what pattern you are talking about... In those lines we replace existing fields with the newly parsed information. In the added line however we unconditionally preserve the original log message.

> Also, since we've failed to have tests for this, it is possible to do some newlogd unit tests for this which would extercise the various combinations of inputs?

I thought about what kind of unit test would have helped us prevent this bug, but I cannot really think of a good format. We cannot simply use static test data, because in case of a change in log format those tests would still pass, giving us the wrong signal.
Also the reasoning behind this change is to prevent loss of information while formatting the logs, which I'm not sure how to test.

Reply:
Sorry, my suggestion was incorrect, which is an indication that the code is hard to understand to begin with.

I wonder whether it would be easier to read if the use of loginfo2 is minimized in the existing code.
`		// Is the Msg itself json?
		logInfo, ok := agentlog.ParseLoginfo(logEntry.Msg)
		if ok {
			// Use the inner JSON struct
			// If the inner has Time or Source set they take
			// precedence over the envelope
			if logInfo.Time == "" {
				logInfo.Time = logEntry.Time
			}
			if logInfo.Source == "" {
				logInfo.Source = logEntry.Source
			}
			// and keep the original message text and fields
			logInfo.Msg = logEntry.Msg
               } else {
			// Start with the envelope
			logInfo.Source = logEntry.Source
			logInfo.Time = logEntry.Time
			logInfo.Msg = logEntry.Msg
			// Some messages have attr=val syntax
			level, timeStr, msg := parseLevelTimeMsg(logInfo.Msg)
			if level != "" {
				logInfo.Level = level
			}
			if timeStr != "" {
				logInfo.Time = timeStr
			}
			if msg != "" {
				logInfo.Msg = msg
			}
		}`

Reply:
:+1: 
@eriknordmark I used your version but changed to comments a little to better describe what's happening